pkg/worker/worker.go

```diff
@@ -0,0 +1,300 @@
+/*
+Package worker allows to start, stop and query Linux processes as concurrent jobs.
+
+Each job is saved to a provided storage via DataStorage interface. Its progress
+is updated each time a new line in StdIn or StdErr appears.
+
+To make use of the package create a new Worker using its constructor - NewWorker
+and provide a correct DataStorage. Control the processes using Start() and Stop()
+methods. Query the output anytime, even when the process is running using Query()
+method.
+
+DataStorage interface must include working data synchronization methods: Lock()
+and Unlock().
+*/
+package worker
+
+import (
+	"io"
+	"log"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/google/uuid"
+)
+
+var (
+	// ErrJobNotFound indicates that the requested job
+	// has not been found in the DataStorage.
+	ErrJobNotFound = errors.New("job has not been found")
+
+	// ErrJobAlreadyStopped indicates that the requested job
+	// has already been stopped.
+	ErrJobAlreadyStopped = errors.New("job has already been stopped")
+
+	// ErrJobAlreadyFinished indicates that the requested job
+	// has already been finished.
+	ErrJobAlreadyFinished = errors.New("job has already been finished")
+)
+
+// DataStorage allows to store and retrieve jobs from any storage that implements the interface.
+// The storage needs to implement sync.Mutex or other form of access synchronization based on
+// Lock() and Unlock() methods.
+type DataStorage interface {
+	Store(job *Job) error
+	Retrieve(uuid uuid.UUID) (*Job, error)
+}
+
+// MemoryStorage is a simple, in-memory implementation of the DataStorage interface.
+// Avoid using in production. Implement DataStorage with database of your choice.
+type MemoryStorage struct {
+	Data  map[uuid.UUID]*Job
```

Comment:
This should still store a non-pointer version of `Job`.
Otherwise, your `Retrieve` returns a pointer to the entry in the map. The caller can mutate it, creating data races.
`Retrieve` should have a return type of `(*Job, error)` to allow returning a nil job, but internally it should make a deep copy of the struct from the map.

Similarly, `Store` should work with non-pointers, so that the caller can't modify the job after storing it.

Alternatively, both `Retrieve` and `Store` should make a manual deep copy of the struct.

Reply:
Ah, I see. I completely misunderstood you before. Yea, this makes sense. The MemoryStore used most of my implementation time because I was running circles with the data races and instead of trying to just store copy, I created mutex-spaghetti without deepcopying.